addons/binding/org.openhab.binding.russound/src/main/java/org/openhab/binding/russound/internal/rio/RioPresetsProtocol.java

```diff
@@ -0,0 +1,476 @@
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.russound.internal.rio;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang.StringUtils;
+import org.openhab.binding.russound.internal.net.SocketSession;
+import org.openhab.binding.russound.internal.net.SocketSessionListener;
+import org.openhab.binding.russound.internal.rio.models.GsonUtilities;
+import org.openhab.binding.russound.internal.rio.models.RioPreset;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * This {@link AbstractRioProtocol} implementation provides the implementation for managing Russound bank presets.
+ * Since refreshing all 36 presets requires 72 calls to russound (for name/valid), we limit how often we can
+ * refresh to {@link #UPDATETIMESPAN}. Presets are tracked by source ID and will only be valid if that source type is a
+ * tuner.
+ *
+ * @author Tim Roberts
+ *
+ */
+
+public class RioPresetsProtocol extends AbstractRioProtocol {
+
+    // logger
+    private Logger logger = LoggerFactory.getLogger(RioPresetsProtocol.class);
+
+    // helper names
+    private final static String PRESET_NAME = "name";
+    private final static String PRESET_VALID = "valid";
+
+    /**
+     * The pattern representing preset notifications
+     */
+    private final Pattern RSP_PRESETNOTIFICATION = Pattern
+            .compile("(?i)^[SN] S\\[(\\d+)\\].B\\[(\\d+)\\].P\\[(\\d+)\\].(\\w+)=\"(.*)\"$");
+
+    /**
+     * The pattern representing a source type notification
+     */
+    private final Pattern RSP_SRCTYPENOTIFICATION = Pattern.compile("^[SN] S\\[(\\d+)\\]\\.type=\"(.*)\"$");
+
+    /**
+     * All 36 presets represented by two dimensions - 8 source by 36 presets
+     */
+    private final RioPreset[][] _presets = new RioPreset[8][36];
+
+    /**
+     * Represents whether the source is a tuner or not
+     */
+    private final boolean[] _isTuner = new boolean[8];
+
+    /**
+     * The {@link Gson} used for JSON operations
+     */
+    private final Gson _gson;
+
+    /**
+     * The {@link ReentrantLock} used to control access to {@link #_lastUpdateTime}
+     */
+    private final Lock _lastUpdateLock = new ReentrantLock();
+
+    /**
+     * The last time the specified source presets were updated via {@link #refreshPresets(Integer)}
+     */
+    private final long[] _lastUpdateTime = new long[8];
+
+    /**
+     * The minimum timespan between updates of source presets via {@link #refreshPresets(Integer)}
+     */
+    private static final long UPDATETIMESPAN = 60000;
```

Comment:
Minor: I think the correct name would be UPDATE_TIME_SPAN.

_Just a suggestion did you know java allows underscore in number to be able to properly display large value, so you could do this:_

```java
private static final long UPDATE_TIME_SPAN_MS = 60_000;
```